scripts/auto-gen.sh

```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
+source "$(dirname $0)/../.ci/variables.sh"
+
+set -ex
+
+add_license() {
+    local FILE="$1"
+    update-license $FILE
+}
+
+export -f add_license
+
+autogen_clear() {
+	local DIR="$1"
+    find $DIR -mindepth 2 -type f -name '*.go' -exec rm -f {} \;
+}
+
+remove_matching_files() {
+    local DIR=$1
+    local FILE_PATTERN=$2
+    find $DIR -type f -name "$FILE_PATTERN" -exec rm -f {} \;
+}
+
+autogen_cleanup() {
+    local DIR="$1"
+    find $DIR -type f -name "*.go" -exec /bin/bash -c 'add_license $0' {} \;
+}
+
+mocks_cleanup_helper() {
+    local FILE=$0
+    local DIR=$(dirname $FILE)
+    add_license $FILE
+
+    # NB(xichen): there is an open issue (https://github.com/golang/mock/issues/30)
+    # with mockgen that causes the generated mock files to have vendored packages
+    # in the import list. For now we are working around it by removing the vendored
+    # path. Also sed -i'' does not work with BSD sed shipped with OS X, whereas
+    # sed -i '' doesn't work with GNU sed, so we work around it by redirecting to a
+    # temp file first and moving it back later.
+    sed "s|$VENDOR_PATH/||" $FILE > $FILE.tmp && mv $FILE.tmp $FILE
+
+    # Strip GOPATH from the source file path
+    sed "s|Source: $GOPATH/src/\(.*\.go\)|Source: \1|" $FILE > $FILE.tmp && mv $FILE.tmp $FILE
+
+    # NB(prateek): running mockclean makes mock-gen idempotent.
+    # NB(xichen): mockclean should be run after the vendor path is stripped.
+    # NB(schallert): mockclean screws up our imports
```

Comment:
lol what does it mess up?

Reply:
Without `mockclean`, `pkg/m3admin/placement/placement_mock.go` is:
```go
package placement

import (
...
placement "github.com/m3db/m3cluster/placement"
)
...
// Get mocks base method
func (m *MockClient) Get() (placement.Placement, error) {
	ret := m.ctrl.Call(m, "Get")
	ret0, _ := ret[0].(placement.Placement)
	ret1, _ := ret[1].(error)
	return ret0, ret1
}
...
```

With `mockclean`, it's
```go
package placement

import (
...
"github.com/m3db/m3cluster/placement"
)
...
// Get mocks base method
func (m *MockClient) Get() (Placement, error) {
	ret := m.ctrl.Call(m, "Get")
	ret0, _ := ret[0].(Placement)
	ret1, _ := ret[1].(error)
	return ret0, ret1
}
```

It tries to import `Placement` as if it's in this package. Guessing it's because base name of this package is the name of this package we're trying to import the type from. Didn't dig in too far past that. This is when tried with both reflect-mode mock generation and file mode.